### ISSUE TITLE
ICU 72.1 wrap doesn't build, not configurable

### ISSUE THREAD

**Issue by srl295** [mesonbuild/wrapdb#950]
Hi! I'm not new to ICU. but I'm pretty new to meson.

Trying to add ICU as a subproject… meson 0.63.0 on mac.

```
icu4c = subproject('icu', default_options: [ 'default_library=static', 'cpp_std=c++17'])

icu_uc = icu4c.get_variable('icuuc_dep')


…
executable(…
   dependencies: [icu_uc],
)
```

Errors below.  Now, is there a way I can set switches on the ICU subproject? 

I'd set something like the following

```
# subprojects/icu/meson.build
add_project_arguments('-DUCONFIG_NO_BREAK_ITERATION=1', '-DUCONFIG_NO_LEGACY_CONVERSION=1', '-DUCONFIG_NO_FORMATTING=1', language: 'cpp')
add_project_arguments('-DU_ATTRIBUTE_DEPRECATED=', language: 'cpp')
```

… but these should be switchable by the caller.

-----

```
[1/417] Compiling C++ object subprojects/icu/source/common/libicuuc.72.dylib.p/ucnvhz.cpp.o
FAILED: subprojects/icu/source/common/libicuuc.72.dylib.p/ucnvhz.cpp.o 
clang++ -arch x86_64 -Isubprojects/icu/source/common/libicuuc.72.dylib.p -Isubprojects/icu/source/common -I../../../subprojects/icu/source/common -fcolor-diagnostics -Wall -Winvalid-pch -Wnon-virtual-dtor -Wextra -Werror -std=c++17 -O3 -DU_COMMON_IMPLEMENTATION -MD -MQ subprojects/icu/source/common/libicuuc.72.dylib.p/ucnvhz.cpp.o -MF subprojects/icu/source/common/libicuuc.72.dylib.p/ucnvhz.cpp.o.d -o subprojects/icu/source/common/libicuuc.72.dylib.p/ucnvhz.cpp.o -c ../../../subprojects/icu/source/common/ucnvhz.cpp
../../../subprojects/icu/source/common/ucnvhz.cpp:558:9: error: 'ucnv_safeClone_72' is deprecated [-Werror,-Wdeprecated-declarations]
        ucnv_safeClone(((UConverterDataHZ*)cnv->extraInfo)->gbConverter, &localClone->subCnv, &size, status);
        ^
../../../subprojects/icu/source/common/unicode/urename.h:706:24: note: expanded from macro 'ucnv_safeClone'
#define ucnv_safeClone U_ICU_ENTRY_POINT_RENAME(ucnv_safeClone)
                       ^
../../../subprojects/icu/source/common/unicode/uvernum.h:123:47: note: expanded from macro 'U_ICU_ENTRY_POINT_RENAME'
#       define U_ICU_ENTRY_POINT_RENAME(x)    U_DEF2_ICU_ENTRY_POINT_RENAME(x,U_ICU_VERSION_SUFFIX)
                                              ^
../../../subprojects/icu/source/common/unicode/uvernum.h:122:51: note: expanded from macro 'U_DEF2_ICU_ENTRY_POINT_RENAME'
#       define U_DEF2_ICU_ENTRY_POINT_RENAME(x,y) U_DEF_ICU_ENTRY_POINT_RENAME(x,y)
                                                  ^
../../../subprojects/icu/source/common/unicode/uvernum.h:121:50: note: expanded from macro 'U_DEF_ICU_ENTRY_POINT_RENAME'
#       define U_DEF_ICU_ENTRY_POINT_RENAME(x,y) x ## y
                                                 ^
<scratch space>:11:1: note: expanded from here
ucnv_safeClone_72
^
../../../subprojects/icu/source/common/unicode/ucnv.h:556:1: note: 'ucnv_safeClone_72' has been explicitly marked deprecated here
U_DEPRECATED UConverter * U_EXPORT2
^
../../../subprojects/icu/source/common/unicode/umachine.h:116:29: note: expanded from macro 'U_DEPRECATED'
#define U_DEPRECATED U_CAPI U_ATTRIBUTE_DEPRECATED
                            ^
../../../subprojects/icu/source/common/unicode/umachine.h:96:52: note: expanded from macro 'U_ATTRIBUTE_DEPRECATED'
#    define U_ATTRIBUTE_DEPRECATED __attribute__ ((deprecated))
                                                   ^
```

**Comment by eli-schwartz** [comment]
> … but these should be switchable by the caller.

I guess you probably want to add `meson_options.txt` in the root of the subproject, containing https://mesonbuild.com/Build-options.html that can be used to toggle things via the command line `meson setup -Dicu:option-name=true` and in the ICU meson.build file:
```meson
if get_option('option-name')
    add_project_arguments('-DFOOBAR=', language: 'cpp')
endif
```

**Comment by srl295** [comment]
@eli-schwartz do you mean structurally a PR to request such a file in https://github.com/mesonbuild/wrapdb/tree/master/subprojects/packagefiles/icu or is there some other way?  As I understand the usage of wrapfiles, I would normally only have `subprojects/icu.wrap` in my own project's source repo

**Comment by srl295** [comment]
ICU has lots of features and switchable options. some of them would disable whole subdirectories, etc.

**Comment by srl295** [comment]
what's the difference between adding items to a new `meson_options.txt` and adding them to https://github.com/mesonbuild/wrapdb/blob/master/subprojects/packagefiles/icu/meson.build  ?

**Comment by srl295** [comment]
ICU's default settings have:

```
# U_ATTRIBUTE_DEPRECATED is defined to hide warnings about deprecated API warnings.                                                                                                                                                
DEFS = -DU_ATTRIBUTE_DEPRECATED=
```

the deprecated warnings are for *callers* of ICU, not internal code. So this seems to be a bug in the wrapfile.

**Comment by srl295** [comment]
When I test against this repo locally in https://github.com/mesonbuild/wrapdb#how-to-test-a-wrap the deprecations are warnings, not errors. So the error part must be coming from my parent project.

I'll open a PR to quell the deprecations…

**Comment by eli-schwartz** [comment]
> do you mean structurally a PR to request such a file in https://github.com/mesonbuild/wrapdb/tree/master/subprojects/packagefiles/icu or is there some other way?

Yes, that would be fantastic. :) Procedurally speaking, this project is fully functional as a meson subproject, so you could copy subprojects/packagefiles/ into another project and patch it... which is like using a private wrap instead, but sharing is caring. :D 

> ICU has lots of features and switchable options. some of them would disable whole subdirectories, etc.

The more closely we come in sync with the upstream build system the better. :) This would both provide feature parity for wrap users, and also serve as a model in case upstream is interested in a meson.build too. So as time permits, people do work on adding new features to wraps. I'd love to review and merge as much or as little as you have the time and resources to work on implementing.

> what's the difference between adding items to a new `meson_options.txt` and adding them to meson.build?

Meson requires options to be in a dedicated file, for a couple reasons. The meson.build cannot programmatically determine what options are available, nor what their values are, they have to be statically defined so that `meson configure` (the equivalent of `./configure --help`) works, and so that repeatedly re-running the configure process cannot cause the value to flip-flop.

(meson has get_option, but not set_option -- programmatic logic should be recalculated afresh on every build based on static inputs like options + environment probing such as dependency lookups or test compiles.)

**Comment by srl295** [comment]
Also is there a way to read a variable from a textfile? I found a way to run an executable, but that seems like too much.

ICU has sources.txt in each subdir:

```
appendable.cpp
bmpset.cpp
brkeng.cpp
brkiter.cpp
```

then meson wouldn't have to hard code:

```
sources = files(
  'appendable.cpp',
  'bmpset.cpp',
  'brkeng.cpp',
  'brkiter.cpp',
```

if you want to get fancy (and actually exceed the upstream's capabilities) there's a dependencies file which tracks which compilation units are needed for which feature.

**Comment by eli-schwartz** [comment]
> Is https://github.com/mesonbuild/icu/tree/67.1 an out of date system?

Indeed, we deprecated the per-wrap repositories when we moved to the current subproject-based monorepo. That old version is imported into the history of this repo.

**Comment by eli-schwartz** [comment]
> Also is there a way to read a variable from a textfile? I found a way to run an executable, but that seems like too much.

Absolutely! I think it's a great idea to share this. You can use:

```meson
fs = import('fs')

sources = fs.read('sources.txt').split()
```

Files that are read in like this during configure are automatically buildsystem reconfigure dependencies.

**Comment by eli-schwartz** [comment]
With a minor tweak to add indentation and single-quoting to sources.txt to align with the meson formatting, I got this diff:
```diff
diff --git a/meson.build b/sources.txt
index d06835fd..20e4abed 100644
--- a/meson.build
+++ b/sources.txt
@@ -1,4 +1,3 @@
-sources = files(
   'appendable.cpp',
   'bmpset.cpp',
   'brkeng.cpp',
@@ -28,15 +27,20 @@ sources = files(
   'icuplug.cpp',
   'loadednormalizer2impl.cpp',
   'localebuilder.cpp',
+  'localematcher.cpp',
+  'localeprioritylist.cpp',
   'locavailable.cpp',
   'locbased.cpp',
   'locdispnames.cpp',
+  'locdistance.cpp',
   'locdspnm.cpp',
   'locid.cpp',
   'loclikely.cpp',
+  'loclikelysubtags.cpp',
   'locmap.cpp',
   'locresdata.cpp',
   'locutil.cpp',
+  'lsr.cpp',
   'lstmbe.cpp',
   'messagepattern.cpp',
   'normalizer2.cpp',
@@ -44,6 +48,7 @@ sources = files(
   'normlzr.cpp',
   'parsepos.cpp',
   'patternprops.cpp',
+  'pluralmap.cpp',
   'propname.cpp',
   'propsvec.cpp',
   'punycode.cpp',
@@ -60,6 +65,7 @@ sources = files(
   'resbund.cpp',
   'resbund_cnv.cpp',
   'resource.cpp',
+  'restrace.cpp',
   'ruleiter.cpp',
   'schriter.cpp',
   'serv.cpp',
@@ -78,6 +84,7 @@ sources = files(
   'ubidi.cpp',
   'ubidi_props.cpp',
   'ubidiln.cpp',
+  'ubiditransform.cpp',
   'ubidiwrt.cpp',
   'ubrk.cpp',
   'ucase.cpp',
@@ -189,170 +196,3 @@ sources = files(
   'uvectr32.cpp',
   'uvectr64.cpp',
   'wintz.cpp',
-)
-
```

Which indicates we seem to be missing a couple things, even.

**Comment by srl295** [comment]
> > Also is there a way to read a variable from a textfile? I found a way to run an executable, but that seems like too much.
> 
> Absolutely! I think it's a great idea to share this. You can use:
> 
> ```meson
> fs = import('fs')
> 
> sources = fs.read('sources.txt').split()
> ```

now y ou're just making it look easy!

> Files that are read in like this during configure are automatically buildsystem reconfigure dependencies.

nice.

**Comment by srl295** [comment]
```
icu| ../subprojects/icu/source/stubdata/meson.build:2: WARNING: Project targets '>=0.55.0' but uses feature introduced in '0.57.0': fs.read.
```

**Comment by eli-schwartz** [comment]
This is a great excuse to bump the minimum version right here: https://github.com/mesonbuild/wrapdb/blob/7a948287954403e774e5a12e0675d0bc220f4213/subprojects/packagefiles/icu/meson.build#L6

The newly updated minimum version is being used for an incredibly valuable new feature that is non-trivial to otherwise do.

We do like to mark the minimum required version, whatever that may be, so that users on older versions get an error message saying they need to upgrade meson, rather than an error message saying "BUG: we don't understand this syntax" or "ERROR: unknown function". ;)

(Also, if it's *easy* to maintain backwards compatibility for people stuck on meson 0.45 because the poor souls are locked into Ubuntu 18.04 LTS, then why not... but that is not the case here.)

**Comment by srl295** [comment]
OK, back to options.

> Meson requires options to be in a dedicated file … (meson_options.txt)

OK, i get it. the options go in meson_options.txt, their implementation in meson.build

However, Is there a way to just generically stuff #defines into a subproject?  I guess what I'd like to do is:

```
icu4c = subproject('icu', default_options: [ 'default_library=static', 'cpp_std=c++17', 'defines+=-DU_CONFIG_THE_THING'])
```

without having to enumerate all possible switches.

**Comment by srl295** [comment]
>  0.45 because the poor souls are locked into Ubuntu 18.04 LTS, then why not... but that is not the case here.

Debian and 0.53 but yeah.

**Comment by neheb** [comment]
Just install the latest version with pip

**Comment by srl295** [comment]
> Just install the latest version with pip

I hear you but it's a downstream issue.

**Comment by neheb** [comment]
wait isn't there a newer version with bullseye-backports?

**Comment by eli-schwartz** [comment]
Perhaps you could define a `feature_flags` array-type option:

```meson
option('feature_flags', type: 'array')
```

And then invoke the project with:
```meson
default_options: ['feature_flags=-DFOO,-DBAR']
```
 or (messy):
```meson
default_options: ['feature_flags=[\'-DFOO\',\'-DBAR\']']
```

(the latter syntax is if you need to pass commas into a value. array-options can be specified on the command line as either comma-separated for convenience, or using painfully quoted but completely unambiguous `['one', 'two', 'three']` style)
